cmd/encryption-v1.go

```diff
@@ -154,64 +70,29 @@ const (
 	// SSESealAlgorithmDareV2HmacSha256 specifies DAREv2 as authenticated en/decryption scheme and SHA256 as cryptographic
 	// hash function for the HMAC PRF.
 	SSESealAlgorithmDareV2HmacSha256 = "DAREv2-HMAC-SHA256"
-
-	// SSEDomain specifies the domain for the derived key - in this case the
-	// key should be used for SSE-C.
-	SSEDomain = "SSE-C"
 )
 
-// hasSSECustomerHeader returns true if the given HTTP header
-// contains server-side-encryption with customer provided key fields.
-func hasSSECustomerHeader(header http.Header) bool {
-	return header.Get(SSECustomerAlgorithm) != "" || header.Get(SSECustomerKey) != "" || header.Get(SSECustomerKeyMD5) != ""
-}
-
-// hasSSECopyCustomerHeader returns true if the given HTTP header
-// contains copy source server-side-encryption with customer provided key fields.
-func hasSSECopyCustomerHeader(header http.Header) bool {
-	return header.Get(SSECopyCustomerAlgorithm) != "" || header.Get(SSECopyCustomerKey) != "" || header.Get(SSECopyCustomerKeyMD5) != ""
+// hasServerSideEncryptionHeader returns true if the given HTTP header
+// contains server-side-encryption.
+func hasServerSideEncryptionHeader(header http.Header) bool {
+	return crypto.S3.IsRequested(header) || crypto.SSEC.IsRequested(header)
 }
 
 // ParseSSECopyCustomerRequest parses the SSE-C header fields of the provided request.
 // It returns the client provided key on success.
-func ParseSSECopyCustomerRequest(r *http.Request) (key []byte, err error) {
+func ParseSSECopyCustomerRequest(r *http.Request, metadata map[string]string) (key []byte, err error) {
 	if !globalIsSSL { // minio only supports HTTP or HTTPS requests not both at the same time
 		// we cannot use r.TLS == nil here because Go's http implementation reflects on
 		// the net.Conn and sets the TLS field of http.Request only if it's an tls.Conn.
 		// Minio uses a BufConn (wrapping a tls.Conn) so the type check within the http package
 		// will always fail -> r.TLS is always nil even for TLS requests.
 		return nil, errInsecureSSERequest
 	}
-	header := r.Header
-	if algorithm := header.Get(SSECopyCustomerAlgorithm); algorithm != SSECustomerAlgorithmAES256 {
-		return nil, errInvalidSSEAlgorithm
-	}
-	if header.Get(SSECopyCustomerKey) == "" {
-		return nil, errMissingSSEKey
+	if crypto.S3.IsEncrypted(metadata) && crypto.SSECopy.IsRequested(r.Header) {
```

Comment:
s/crypto.S3.IsEncrypted/crypto.S3.IsRequested(r.Header) and remove metadata parameter.

Reply:
@aead, for sse-s3 encrypted objects on the source side, the  header would not be present. We will need to make it out from the metadata. Hence this check is needed

Reply:
@poornas - I tested this using
```
client, err := minio.NewV4("s3.amazonaws.com", "access-key", "secret-key", true)
	if err != nil {
		log.Fatalln(err)
	}

	bucket, object := "aead", "object"
	data := []byte("Hello World")

	_, err = client.PutObject(bucket, object, bytes.NewReader(data), int64(len(data)), minio.PutObjectOptions{ServerSideEncryption: encrypt.NewSSE()})
	if err != nil {
		log.Fatalln(err)
	}

	src := minio.NewSourceInfo(bucket, object, nil)
	dst, _ := minio.NewDestinationInfo(bucket, object+"copy", encrypt.DefaultPBKDF([]byte("123456"), nil), nil)

	if err = client.CopyObject(dst, src); err != nil {
		log.Fatalln(err)
	}
```

AWS accepts it and IMO this is correct. If the object is encrypted using SSE-C you it can't be SSE-S3 and if the object is encrypted using SSE-C it cannot be SSE-S3.
If the object is not encrypted (SSE-C) but client provides SSE-C copy headers AWS fails with `The encryption parameters are not applicable to this object.` But this should be generic behavior for all non-encrypted objects

Reply:
In essence, this check can be left as is.